server/src/main/java/org/elasticsearch/action/support/FanOutListener.java

```diff
@@ -0,0 +1,294 @@
+     * Listeners added strictly before this listener is completed will themselves be completed in the order in which their subscriptions
+     * were received. However, there are no guarantees about the ordering of the completions of listeners which are added concurrently with
+     * (or after) the completion of this listener.
+     * <p>
+     * If the subscribed listener is not completed immediately then it will be completed on the thread, and in the {@link ThreadContext}, of
+     * the thread which completes this listener.
+     */
+    public final void addListener(ActionListener<T> listener) {
+        addListener(listener, EsExecutors.DIRECT_EXECUTOR_SERVICE, null);
+    }
+
+    /**
+     * Add a listener to this listener's collection of subscribers. If this listener is complete, this method completes the subscribing
+     * listener immediately with the result with which this listener was completed. Otherwise, the subscribing listener is retained and
+     * completed when this listener is completed.
+     * <p>
+     * Subscribed listeners must not throw any exceptions. Use {@link ActionListener#wrap(ActionListener)} if you have a listener for which
+     * exceptions from its {@link ActionListener#onResponse} method should be handled by its own {@link ActionListener#onFailure} method.
+     * <p>
+     * Listeners added strictly before this listener is completed will themselves be completed in the order in which their subscriptions
+     * were received. However, there are no guarantees about the ordering of the completions of listeners which are added concurrently with
+     * (or after) the completion of this listener.
+     *
+     * @param executor      If not {@link EsExecutors#DIRECT_EXECUTOR_SERVICE}, and the subscribing listener is not completed immediately,
+     *                      then it will be completed using the given executor. If the subscribing listener is completed immediately then
+     *                      this completion happens on the subscribing thread.
+     * @param threadContext If not {@code null}, and the subscribing listener is not completed immediately, then it will be completed in
+     *                      the given thread context. If {@code null}, and the subscribing listener is not completed immediately, then it
+     *                      will be completed in the {@link ThreadContext} of the completing thread. If the subscribing listener is
+     *                      completed immediately then this completion happens in the {@link ThreadContext} of the subscribing thread.
+     */
+    @SuppressWarnings({ "rawtypes" })
+    public final void addListener(ActionListener<T> listener, Executor executor, @Nullable ThreadContext threadContext) {
+        if (tryComplete(ref.get(), listener)) {
+            return;
+        }
+
+        final var wrappedListener = fork(executor, preserveContext(threadContext, listener));
+        var currentValue = ref.compareAndExchange(EMPTY, wrappedListener);
```

Comment:
I think it would be helpful to have a few of these vars changed to the actual types - would help to make it clear when a variable can be any type & when its a specific type